Add drbd module and state.

nick-wang opened the pull request.

nick-wang commented:
@arbulu89 I have fixed some of the issue in codeclimate. However, still some of the issues report of the latest code [codeclimate](https://codeclimate.com/github/SUSE/salt-saphana/pull/17). 
I don't think fix them make the code easier to understand/maintain, since code in upstream share the same style. What 's your opinion?

arbulu89 commented:
> @arbulu89 I have fixed some of the issue in codeclimate. However, still some of the issues report of the latest code [codeclimate](https://codeclimate.com/github/SUSE/salt-saphana/pull/17).
> I don't think fix them make the code easier to understand/maintain, since code in upstream share the same style. What 's your opinion?

Hi @nick-wang,
Don't worry about them. Codeclimate complains about everything, and most salt related code always will have some issues

diegoakechi commented:
Ok, it was too good to be true... @nick-wang can you rebase your branch.. we did the changes to rename the repo and it broke the changes you did on the RPM package.

diegoakechi commented on salt-saphana.spec:
```diff
@@ -30,6 +32,9 @@ Source0:        %{name}-%{version}.tar.gz
 BuildRoot:      %{_tmppath}/%{name}-%{version}-build
 BuildArch:      noarch
 Requires:       %{python_module shaptools}
+%if %{with drbd}
```
@nick-wang @arbulu89 maybe not needed for this initial PR, but would make sense to me to split this into subpackages, like:

- salt-shaptools-crm
- salt-shaptools-hana
- salt-shaptoos-drdb

where each subpackage installs its modules and states...

nick-wang replied:
> Ok, it was too good to be true... @nick-wang can you rebase your branch.. we did the changes to rename the repo and it broke the changes you did on the RPM package.

Sure, will do it tomorrow.
Thanks a lot for the change. I know fix all the dependencies in all repos/pkgs is not as easy as a simple `sed`

nick-wang replied:
Sounds good/correct to me for the split. 
And i think it is better to have conditions for all subpackages, even for crm and hana. Because if only parts of the module in package merged into upstream, it is easier for us to disabled the build/installation of specific subpackages only.

Anyway, I will remove the condition of `drbd` during the rebase and decide in future.

nick-wang commented:
> Ok, it was too good to be true... @nick-wang can you rebase your branch.. we did the changes to rename the repo and it broke the changes you did on the RPM package.

Sure, will do it tomorrow.
Thanks a lot for the change. I know fix all the dependencies in all repos/pkgs is not as easy as a simple `sed`

nick-wang commented:
> Ok, it was too good to be true... @nick-wang can you rebase your branch.. we did the changes to rename the repo and it broke the changes you did on the RPM package.

Finished rebase on the latest change.

diegoakechi commented:
@nick-wang Can you include the missing coverage for DRDB? If is it too complex, we can accept this way, but if not, I think it worth to do the improvement before merge.

nick-wang commented:
> @nick-wang Can you include the missing coverage for DRDB? If is it too complex, we can accept this way, but if not, I think it worth to do the improvement before merge.

Will try~ I was trying to find where exactly 3% it point to, but i didn't find a report with details:( Do you have any idea on where i can see it?

nick-wang commented:
Enhance test coverage to 100%

arbulu89 commented on tests/unit/modules/test_drbd.py:
```diff
@@ -0,0 +1,648 @@
+'''
+    :codeauthor: Jayesh Kariya <jayeshk@saltstack.com>
+'''
+
```
In general I prefer to have one test case for each scenario. In this case, you have all of the scenarios of a method in one test (what you call subtest). I prefer the other way to avoid the mock resets and so on.
Anyway, it is OK. Just I wanted to comment it. Usually I prefer to have the same way of doing tests into the same project hehe.

nick-wang replied:
I don't change this so far.

I am using subtest instead of new functions mainly for two reason:
1. Match one test case to one function without adding more functions.
2. Different subtest can share same variable if possible, though not many cases. 
For example, mock the time.time() method can be reused in many cases, though need deepcopy() to duplicate.
Or using generic mock_cmd_true() and mock_cmd_false() for most cases, though i don't really do like that.

nick-wang replied:
I just found copy.deepcopy() can't pass the test of python2... Only work in python3. Refer to [travis-log](https://travis-ci.org/SUSE/salt-shaptools/builds/528682932)

I will fix it in the later commit.

nick-wang replied:
it seems a py2 issue for attributes can't be pickled, only fixed in python3.2.
So will define multiple instances instead of deepcopy for compatibility.

nick-wang replied:
Fixed in 81287f072540ab7d713af09d0a88363e42d94079

arbulu89 commented on tests/unit/modules/test_drbd.py:
```diff
@@ -0,0 +1,648 @@
+# -*- coding: utf-8 -*-
+'''
+    :codeauthor: Jayesh Kariya <jayeshk@saltstack.com>
```
???

nick-wang replied:
This is just inherit from salt upstream. drbd is exist for a long time, but only one function and will not support in the latest drbd tools.

I think i need to leave as it is, just add me in the list.

arbulu89 replied:
Thanks for clarifying!

arbulu89 commented on tests/unit/states/test_drbd.py:
```diff
@@ -0,0 +1,907 @@
+            'comment': 'Resource {} not defined in your config.'.format(RES_NAME),
+        }
+
+        mock = MagicMock(return_value=1)
```
Use a more representative name

nick-wang replied:
Fixed.

arbulu89 commented on tests/unit/states/test_drbd.py:
```diff
@@ -0,0 +1,907 @@
+        mock = MagicMock(return_value=1)
+
+        with patch.dict(drbd.__salt__, {'cmd.retcode': mock}):
+            self.assertDictEqual(drbd.initialized(RES_NAME), ret)
```
Use plain asserts

nick-wang replied:
Fixed.

arbulu89 commented on tests/unit/states/test_drbd.py:
```diff
@@ -0,0 +1,907 @@
+        mock_status = MagicMock(return_value=res_status)
+        mock_sync_status = MagicMock(return_value=0)
+
+        with patch.dict(drbd.__opts__, {'test': True}):
```
You can combine all of this patches to work together. Check test cases in crm and hana

nick-wang replied:
Got it, fixed.

arbulu89 commented on tests/unit/states/test_drbd.py:
```diff
@@ -0,0 +1,907 @@
+                    self.assertDictEqual(drbd.demoted(RES_NAME), ret)
+                    mock_secondary.assert_called_once_with(name=RES_NAME)
+
+    def test_wait_for_successful_synced(self):
```
This method uses the `time` and `sleep` methods, right? We should mock it, I wounldn't like to have time dependant unit tests

nick-wang replied:
Got it. Fixed in the latest commit.

arbulu89 commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+        while True:
+
+            if time.time() > starttime + timeout:
+                LOG.error('Syncing of {} is not synced within {}s.'.format(
```
Is this error message needed? You are printing the same in the salt comment

nick-wang replied:
Remove the error log with result['comment'] only

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+# -*- coding: utf-8 -*-
+'''
+DRBD administration module
```
Parts of drbd module exists/inherit in the upstream(though most codes are from me...), so i am not sure whether suitable to name me as the maintainer.

But anyway, i will list there unless anyone object:)

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+
+import salt.utils.json
+
+LOG = logging.getLogger(__name__)
```
No problem

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+
+# Define the module's virtual name
+__virtualname__ = 'drbd'
+
```
__virtual__ method is not a must. Few modules are created without __virtual__ method, drbd.py in upstream is one of them, though no harm to have one.  And the __virtualname__ variable exist for doc build system.

But like you said, no harm to have one, so i will add one.

arbulu89 replied:
Perfect! No worries.
I use the __virtual__ method to make sure if the needed utiliities are installed (drbdadm in this case). If it's not installed it raises an error and it is easier to know why the module was not loaded

nick-wang replied:
Fixed in the latest version

arbulu89 replied:
Perfect! Thx!

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+
+def _analyse_overview_field(content):
+    '''
+    Split the field in drbd-overview
```
Yes, it is intend to return the same content twice, since the input of this should like "Connected(2*)". (Only two nodes supported in that version)

In the latest drbd version, it function is no longer exist. Leave here just for compatible of old version and upstream.

arbulu89 replied:
Ok. So would be nice just to comment that the method is deprecated since X version

nick-wang replied:
Was mentioned in the function `__doc__` of `overview()`

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    Count the number of spaces before the first character
+    '''
+    if line.split('#')[0].strip() == "":
+        return None
```
Actually, this is not an exception but represent not a proper line to parser for drbd.
In this two lines, means do nothing with the empty line or a comment.

The return number is value for `space` startswith, 0 is also valid.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    Check whether all local volumes are UpToDate.
+    '''
+
+    ret = False
```
Removed.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+
+    result = __salt__['cmd.run_all'](cmd)
+    if result['retcode'] != 0:
+        LOG.info('No status due to {} ({}).'.format(result['stderr'], result['retcode']))
```
fixed.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    #    volume:0 peer-disk:Inconsistent resync-suspended:peer
+    #    volume:1 peer-disk:Inconsistent resync-suspended:peer
+
+    result = __salt__['cmd.run_all'](cmd)
```
Yes, it works~ I tested before.

nick-wang replied:
Change to **string** for an unified style.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+        salt '*' drbd.adjust name=<resource name>
+    '''
+
+    ret = []
```
Removed.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    cmd = ['drbdsetup', 'show']
+    cmd.append(name)
+
+    if json:
```
Yes, only support `json` so far as a drbd native format of drbdsetup. Otherwise, use `status()` instead at the moment.

arbulu89 replied:
Ok! Maybe we should just log a warning/info message

nick-wang replied:
Mentioned in `__doc__`. Do you think we still need a warning/info message?

arbulu89 replied:
In my opinion, if only json is accepted, we should even remove that option. If the idea is to add more formats in the future, we could handle that there. Anyway, I would appreciate a warning message but it's not totally mandatory anyway

nick-wang replied:
Got your idea.
I will remove the `json` condition but leave the description in `__doc__`.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    cmd = ['drbdsetup', 'status']
+    cmd.append(name)
+
+    if json:
```
Yes, only support `json` so far as a drbd native format of drbdsetup. Otherwise, use `status()` instead at the moment.

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+from salt.ext import six
+import time
+
+LOG = logging.getLogger(__name__)
```
Fixed

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+LOG = logging.getLogger(__name__)
+
+__virtualname__ = 'drbd'
+
```
Fixed.

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+
+
+def _resource_not_exist(name):
+    cmd = 'drbdadm dump {}'.format(name)
```
Both list and string are supported after the test.

Personally i like string, but the upstream drbd use list, so i just leave as it it in module. I will convert them to list.

nick-wang replied:
I will change all cmd to string.

arbulu89 replied:
Ok thanks! I like to have same patterns in that cases and if the change doesn't request much effort I would appreciate it

nick-wang replied:
All fixed.

arbulu89 replied:
Perfect! Thanks!

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+    cmd = 'drbdadm dump {}'.format(name)
+    result = __salt__['cmd.retcode'](cmd)
+
+    if result != 0:
```
Fixed

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+    res = _get_res_status(name)
+    if not res:
+        ret['result'] = True
+        ret['comment'] = 'Resource {} has already stopped.'.format(name)
```
Fixed.

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+    res = _get_res_status(name)
+    if res:
+        ret['result'] = True
+        ret['comment'] = 'Resource {} has already started.'.format(name)
```
Fixed.

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+def _get_res_status(name):
+    try:
+        result = __salt__['drbd.status'](name=name)
+    except Exception:
```
Change the generic Exception to specific CommandExecutionError.

nick-wang commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,450 @@
+        return ret
+
+
+def wait_for_successful_synced(name, interval=30, timeout=600, **kwargs):
```
I understand the drawback of using kwargs, but i think it is suitable in this case, though not fully implemented yet.

The drbd resource running status could be a big matrix and quite flexible. The idea is compare the dynamic input with the status `json` output based on the configuration. It is more than 50 parameters, also mutable and in development, so i think it is better to leave **kwargs. And considering the function is a monitor only without real action, so no harm if the input is wrong.

However, in this version, i only implement the local/peer updated as a minimal set for the formula.

arbulu89 replied:
Ok, if this is the use case I understand why kwargs is used. No problem in that case

nick-wang replied:
Thanks!

nick-wang commented on tests/unit/modules/test_drbd.py:
```diff
@@ -0,0 +1,648 @@
+        mock = MagicMock(return_value='Salt:Stack True master/minion \
+        UpToDate/UpToDate True None 50 50 666 888')
+        with patch.dict(drbd.__salt__, {'cmd.run': mock}):
+            self.assertDictEqual(drbd.overview(), ret)
```
ok. I don't like it either, since it doesn't show the type of the parameter.
But i will respect and follow. Quite many need to be changed...

arbulu89 replied:
Yes...
Anyway, this was told to me when I created the PR with the other hana/crm modules. However I have found several modules using new assert methods. So, the rules are not followed by everyone...

nick-wang commented on tests/unit/modules/test_drbd.py:
```diff
@@ -0,0 +1,648 @@
+        '''
+        Test if createmd function work well
+        '''
+        mock = MagicMock(return_value=True)
```
Sure, fixed

nick-wang commented on tests/unit/modules/test_drbd.py:
```diff
@@ -0,0 +1,648 @@
+
+        mock = MagicMock(return_value=fake)
+
+        with patch.dict(drbd.__salt__, {'cmd.run_all': mock}):
```
Sorry i don't get your idea. Do you mean 'cmd.run_all' and 'cmd.run' or the way of using patch?

nick-wang commented:
@arbulu89 Appreciate your valuable idea. This is a summary of my fixes based on your comments to ease the review.
I fixed most of the comments, except:
1. Remove the `global` keyword by salt built-in `__salt__['drbd.resource']`: This was failed in my previous test due to different behavior, i will dig more to see the possibility of replacement.
2. Remain using subcases instead of different functions. Refer to comment above for details. (I need to fix the incompatible of deepcopy in python2 later)
3. For the two functions that support `json` format only. So far, i only mentioned it in function's `__doc__`, do you think it is better to add a warning/error to log when calling without `json`?

Thanks a lot for the review. It is really helpful:)

nick-wang commented:
> * Remove the `global` keyword by salt built-in `__salt__['drbd.resource']`: This was failed in my previous test due to different behavior, i will dig more to see the possibility of replacement.
  Fixed in 7a7ebc2c70efeac8cc15a098f861ebe499d2758d via ``__context__``
> * Remain using subcases instead of different functions. Refer to comment above for details. (I need to fix the incompatible of deepcopy in python2 later)
  Fixed in 81287f072540ab7d713af09d0a88363e42d94079 by creating multiple instances.

@arbulu89 I think i fixed all the spots that you comment, please help to review. Thanks in advanced.

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+
+
+# Global para for func status
+ret = []
```
Fixed in 7a7ebc2c70efeac8cc15a098f861ebe499d2758d

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    '''
+    Analyse the line of peer nodes of ``drbdadm status``
+    '''
+    global lastpnodevolumes
```
Fixed in 7a7ebc2c70efeac8cc15a098f861ebe499d2758d

nick-wang commented on salt/modules/drbd.py:
```diff
@@ -0,0 +1,634 @@
+    '''
+    Analyse the line of local resource of ``drbdadm status``
+    '''
+    global resource
```
Fixed in 7a7ebc2c70efeac8cc15a098f861ebe499d2758d

nick-wang commented:
@arbulu89 @diegoakechi Thanks a lot for the review and your opinions. They are valuable and make the code better.

# I will create a new pr later for removing the `json` condition, since it is the only support method in the near plan.

nick-wang commented:
> Actually, I didn't know this `__context__` usage, but maybe it fits better than using `__salt__`

Me too. I just learned and referenced the [__context__ usage](https://docs.saltstack.com/en/latest/topics/development/modules/developing.html#context) today, some modules use it to remove global variable.
